src/OpenDiffix.Core/Anonymizer.fs

```diff
@@ -47,25 +47,39 @@ let private generateNoise salt stepName stdDev noiseLayers =
 
 // Compacts flattening intervals to fit into the total count of contributors.
 // Both intervals are reduced proportionally, with `topCount` taking priority.
+// NOTE `totalCount >= outlierCount.Lower + topCount.Lower` is assumed
 let private compactFlatteningIntervals outlierCount topCount totalCount =
   let totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount
 
   if totalAdjustment <= 0 then
     outlierCount, topCount // no adjustment needed
   else
+    // NOTE: at this point we know `0 <= totalAdjustment <= outlierRange + topRange` (*)
+    //       because `totalAdjustment = outlierCount.Upper + topCount.Upper - totalCount
+    //                               <= outlierCount.Upper + topCount.Upper - outlierCount.Lower - topCount.Lower`
+    //       by the condition checked in `aidFlattening`
     let outlierRange = outlierCount.Upper - outlierCount.Lower
-
-    if outlierRange > topCount.Upper - topCount.Lower then
-      failwith "Invalid config: OutlierCount interval is larger than TopCount interval."
-
-    let outlierAdjustment, topAdjustment =
-      if outlierRange < totalAdjustment / 2 then
-        outlierRange, totalAdjustment - outlierRange
-      else
-        totalAdjustment / 2, totalAdjustment - totalAdjustment / 2
-
-    { outlierCount with Upper = outlierCount.Upper - outlierAdjustment },
-    { topCount with Upper = topCount.Upper - topAdjustment }
+    let topRange = topCount.Upper - topCount.Lower
+    // `topAdjustment` will be half of `totalAdjustment` rounded up, so it takes priority as it should
+    let outlierAdjustment = totalAdjustment / 2
+    let topAdjustment = totalAdjustment - outlierAdjustment
+
+    if outlierRange >= outlierAdjustment && topRange >= topAdjustment then
+      // both ranges are compacted at same rate
+      { outlierCount with Upper = outlierCount.Upper - outlierAdjustment },
+      { topCount with Upper = topCount.Upper - topAdjustment }
+    elif outlierRange < outlierAdjustment then
+      // `outlierCount` is compacted as much as possible by `outlierRange`, `topCount` takes the surplus adjustment
+      { outlierCount with Upper = outlierCount.Lower },
+      { topCount with Upper = topCount.Upper - totalAdjustment + outlierRange }
+    elif topRange < topAdjustment then
+      // vice versa
+      { outlierCount with Upper = outlierCount.Upper - totalAdjustment + topRange },
+      { topCount with Upper = topCount.Lower }
+    else
+      // Not possible. Otherwise `outlierRange + topRange < outlierAdjustment + topAdjustment = totalAdjustment` but we
+      // knew the opposite was true in (*) above
+      failwith "Not possible"
```

Comment:
Should be a better error message. Or is this 100% sure to never run?

Reply:
In cases like this, I find it better to omit the last `if` branch and instead assert the condition is true in the `else` branch.

Reply:
> Should be a better error message. Or is this 100% sure to never run?

So the comments aim at "proving" that 100% won't ever run. What would be better message for a runtime assertion failure like this? Do we have a precedent?

re the if/elif/else suggestion, I'll take a look, together with the proof/assertion suggestions

Reply:
> In cases like this, I find it better to omit the last if

OK, I see where the confusion came from, this is a different case i think. We have 2x2=4 mutually exclusive conditions, depending on which adjustment "fits" the range. I'll refactor so this stands out better